ee/session_recordings/ai/output_data.py

```diff
@@ -0,0 +1,163 @@
+    event = serializers.CharField(min_length=1, max_length=128)
+    event_type = serializers.CharField(max_length=128, required=False, allow_null=True)
+
+
+class RawSessionSummarySerializer(serializers.Serializer):
+    """
+    Raw session summary coming from LLM.
+    """
+
+    summary = serializers.CharField(min_length=1, max_length=2048)
+    key_events = serializers.ListField(child=RawKeyEventSerializer(), allow_empty=False)
+
+
+class SessionSummarySerializer(serializers.Serializer):
+    """
+    Session summary enriched with metadata.
+    """
+
+    summary = serializers.CharField(min_length=1, max_length=2048)
+    key_events = serializers.ListField(child=KeyEventSerializer(), allow_empty=False)
+
+
+def load_raw_session_summary_from_llm_content(
+    llm_response: ChatCompletion, allowed_event_ids: list[str], session_id: str
+) -> RawSessionSummarySerializer:
+    if not llm_response.choices or not llm_response.choices[0].message or not llm_response.choices[0].message.content:
+        raise ValueError(f"No LLM content found when summarizing session_id {session_id}: {llm_response}")
+    raw_content: str = llm_response.choices[0].message.content
+    try:
+        # Strip the first and the last line of the content to load the YAML data only into JSON
+        # TODO Work on a more robust solution
+        json_content: dict = yaml.safe_load(raw_content.strip("```yaml\n").strip("```").strip())  # noqa: B005
+    except Exception as e:
+        raise ValueError(f"Error loading YAML content into JSON when summarizing session_id {session_id}: {e}")
+    # Validate the LLM output against the schema
+    raw_session_summary = RawSessionSummarySerializer(data=json_content)
+    if not raw_session_summary.is_valid():
+        raise ValueError(
+            f"Error validating LLM output against the schema when summarizing session_id {session_id}: {raw_session_summary.errors}"
+        )
+    # Ensure that LLM didn't hallucinate events
+    for key_event in raw_session_summary.data["key_events"]:
+        event_id = key_event.get("event_id")
+        if not event_id:
+            raise ValueError(
+                f"LLM returned event without event_id when summarizing session_id {session_id}: {raw_session_summary.data}"
+            )
+        if key_event["event_id"] not in allowed_event_ids:
+            raise ValueError(
+                f"LLM hallucinated event_id {event_id} when summarizing session_id "
+                f"{session_id}: {raw_session_summary.data}"
+            )
+    return raw_session_summary
+
+
+# TODO Rework the logic, so events before the recording are marked as "LOAD", not 00:00
+def calculate_time_since_start(session_timestamp: str, session_start_time: datetime | None) -> int | None:
+    if not session_start_time or not session_timestamp:
+        return None
+    timestamp_datetime = datetime.fromisoformat(session_timestamp)
+    # TODO Check why the event could happen before the session started
```

Comment:
there are lots of ways we can get an event session that doesn't overlap a recording session
you can send events immediately on SDK boot but we have to download the rrweb js before recording starts for example

or you might have an event (or other type of) trigger before we start recording